File: project4_Healthcare_Analytics_Support_Code/python/src/ipinstance.py
from dataclasses import dataclass
import math
import numpy  as np
from docplex.mp.model import Model
import heapq

@dataclass(frozen=True)
class IPConfig:
   numTests: int # number of tests
   numDiseases : int # number of diseases
   costOfTest: np.ndarray #[numTests] the cost of each test
   A : np.ndarray #[numTests][numDiseases] 0/1 matrix if test is positive for disease


#  * File Format
#  * #Tests (i.e., n)
#  * #Diseases (i.e., m)
#  * Cost_1 Cost_2 . . . Cost_n
#  * A(1,1) A(1,2) . . . A(1, m)
#  * A(2,1) A(2,2) . . . A(2, m)
#  * . . . . . . . . . . . . . .
#  * A(n,1) A(n,2) . . . A(n, m)


# each column must be unique from the other columns
# select tests such that each column is unique and minimize the cost
# you can tell if columns are unique by checking if the dot product of the columns is 0 
        # sum of absolute value of differences is 0

def data_parse(filename : str) :
    try:
      with open(filename,"r") as fl:
        numTests = int(fl.readline().strip()) #n 
        numDiseases = int(fl.readline().strip()) #m

        costOfTest = np.array([float(i) for i in fl.readline().strip().split()])

        A = np.zeros((numTests,numDiseases))
        for i in range(0,numTests):
          A[i,:] = np.array([int(i) for i in fl.readline().strip().split() ])
        return numTests,numDiseases,costOfTest,A
    except Exception as e:
       print(f"Error reading instance file. File format may be incorrect.{e}")
       exit(1)

class IPInstance:

  def __init__(self,filename : str) -> None:
    numT,numD,cst,A = data_parse(filename)
    self.numTests = numT
    self.numDiseases = numD
    self.costOfTest = cst
    self.A = A
    self.model = Model() #CPLEX solver
    self.difference_matrix = self.create_difference_matrix()
    self.objVal, self.solution = self.solve()
    self.constraint_map = self.preComputeConstraints()
  
  def create_difference_matrix(self):
    difference_matrix = np.zeros((self.numTests, self.numDiseases * (self.numDiseases - 1) // 2), dtype=int)

    col_index = 0
    for j in range(self.numDiseases):
        for k in range(j + 1, self.numDiseases):
            # XOR each element of the two disease columns and store the result in the difference matrix
            difference_matrix[:, col_index] = np.bitwise_xor(self.A[:, j].astype(int), self.A[:, k].astype(int))
            col_index += 1

    return difference_matrix
  
  def solve(self):
    # 1. Define decision variables
    # x = self.model.integer_var_list(self.numTests, 0, 1, name="x")
    x = self.model.continuous_var_list(self.numTests, 0, 1, name="x")

    # 2. Set objective function (minimize total cost)
    self.model.minimize(self.model.sum(self.costOfTest[i] * x[i] for i in range(self.numTests)))

    # 3. Add differentiation constraints for each disease pair
    for j in range(self.difference_matrix.shape[1]):
      self.model.add_constraint(self.model.sum(x[i] * self.difference_matrix[i, j] for i in range(self.difference_matrix.shape[0])) >= 1)
    
    # 4. Solve the model
    self.model.solve()

    # solution = self.model.solution.get_values(x)
    # print("Solution:", solution)
    # print("Objective value:", self.model.solution.get_objective_value())
    return self.model.solution.get_objective_value(), self.model.solution.get_values(x)
  
  def apply_constraints(self, constraints: dict[int, int]):
    docplex_constraints = []
    for key, val in constraints.items():
    #   if val == 1:
    #     c = self.model.get_var_by_name("x_" + str(key)) >= 1
    #   else:
    #     c = self.model.get_var_by_name("x_" + str(key)) <= 0
    #   docplex_constraints.append(c)
      docplex_constraints.append(self.constraint_map["x_" + str(key)][val])
    self.model.add_constraints(docplex_constraints)
    self.model.solve()

    new_sol = []
    ret_val = math.inf
    if self.model.solution is not None:
      new_sol = [self.model.solution.get_value(var) for var in self.model.iter_variables()]
      # new_sol = self.model.get_all_values(self.model.iter_variables())
      ret_val = self.getObjectiveValue(new_sol)
    # Clear the constraints
    self.model.remove_constraints(docplex_constraints)

    return new_sol, ret_val


  def preComputeConstraints(self):
    constraint_map = dict()
    for i in range(self.numTests):
      constraint_map["x_" + str(i)] = [None]*2
      c = self.model.get_var_by_name("x_" + str(i)) <= 0
      constraint_map["x_" + str(i)][0] = c
      c = self.model.get_var_by_name("x_" + str(i)) >= 1
      constraint_map["x_" + str(i)][1] = c
    return constraint_map
  
  def branch_and_bound(self):
    upper_bound = math.inf
    lower_bound = self.objVal
    candidate_solutions = []
    candidate_solutions.append((self.solution, self.model, self.objVal))
    best_solution = []
    while candidate_solutions:
      current_solution, current_model, cur_val = candidate_solutions.pop()
      if cur_val > upper_bound:
        continue
      branch_idx = self.greatest_fractional_idx(current_solution)
      if branch_idx == -1:
        new_val = self.getObjectiveValue(current_solution)
        print("integer solution found:", new_val)
        if new_val < upper_bound:
          upper_bound = new_val
          best_solution = current_solution
        continue
      else:
        new_model = current_model.copy()
        new_model.add_constraint(new_model.get_var_by_name("x_" + str(branch_idx)) >= 1)
        new_model.solve()
        new_solution = []
        if new_model.solution != None:
          for var in new_model.iter_variables():
            new_solution.append(new_model.solution.get_value(var))
          new_val = self.getObjectiveValue(new_solution)
          if new_val < upper_bound:
            candidate_solutions.append((new_solution, new_model, new_val))


        new_model = current_model.copy()
        new_model.add_constraint(new_model.get_var_by_name("x_" + str(branch_idx)) <= 0)
        new_model.solve()
        new_solution = []
        if new_model.solution != None:
          for var in new_model.iter_variables():
            new_solution.append(new_model.solution.get_value(var))
          new_val = self.getObjectiveValue(new_solution)
          if new_val < upper_bound:
            candidate_solutions.append((new_solution, new_model, new_val))

    return upper_bound, best_solution


  
  def branch_and_bound_dfs(self):
    upper_bound = math.inf
    best_solution = []

    def dfs(current_solution, constraints, cur_val):
      nonlocal upper_bound, best_solution

      if cur_val >= upper_bound:
        return  # Prune branch

      branch_idx = self.greatest_fractional_idx(current_solution)
      if branch_idx == -1:
        print("integer solution found:", cur_val)
        if cur_val < upper_bound:
          upper_bound = cur_val
          best_solution = current_solution
        return  # Reached leaf node

      # Explore branches (setting x_i to 1 and 0)
      for branch_value in [1, 0]:
        new_constraints = constraints.copy()
        new_constraints[branch_idx] = branch_value
        new_solution, new_obj_val = self.apply_constraints(new_constraints)
        if new_obj_val < upper_bound:
          dfs(new_solution, new_constraints, new_obj_val)

    # Start the DFS from the initial solution
    initial_solution = self.solution
    initial_obj_val = self.objVal
    dfs(initial_solution, dict(), initial_obj_val)

    return upper_bound, best_solution

  def branch_and_bound_best(self):
    upper_bound = math.inf
    best_solution = []
    # Use a priority queue (min-heap)
    priority_queue = [(self.objVal, self.solution, dict())]  # (priority, solution, constraints)
    heapq.heapify(priority_queue)

    while priority_queue:
      new_val, current_solution, constraints = heapq.heappop(priority_queue)
      if new_val > upper_bound:
        continue
      branch_idx = self.greatest_fractional_idx(current_solution)
      if branch_idx == -1:
        print("integer solution found:", new_val)
        if new_val < upper_bound:
          upper_bound = new_val
          best_solution = current_solution
      else:
        new_constraints = constraints.copy()
        new_constraints[branch_idx] = 1
        new_solution, new_obj_val = self.apply_constraints(new_constraints)
        upper_bound = min(upper_bound, self.getRoundUpObjectiveValue(new_solution))
        if new_obj_val <= upper_bound:
          heapq.heappush(priority_queue, (new_obj_val, new_solution, new_constraints))

        new_constraints = constraints.copy()
        new_constraints[branch_idx] = 0
        new_solution, new_obj_val = self.apply_constraints(new_constraints)
        upper_bound = min(upper_bound, self.getRoundUpObjectiveValue(new_solution))
        if new_obj_val <= upper_bound:
          heapq.heappush(priority_queue, (new_obj_val, new_solution, new_constraints))
        
    return upper_bound, best_solution


  def branch_idx(self, cur_sol):
    for i in range(self.numTests):
      if cur_sol[i] > 0 and cur_sol[i] < 1.0:
        return i
    return -1


  def greatest_fractional_idx(self, arr):
    max_fractional = 0
    max_idx = -1
    for i, num in enumerate(arr):
      fractional_part = abs(num - int(num))  # Get the absolute value of the fractional part
      if fractional_part > max_fractional:
        max_fractional = fractional_part
        max_idx = i
    return max_idx

  def distance_to_half(self, arr):
    max_fractional = 0
    max_idx = -1
    for i, num in enumerate(arr):
      fractional_part = abs(num - 0.5)  # Get the absolute value of the fractional part
      if fractional_part > max_fractional:
        max_fractional = fractional_part
        max_idx = i
    return max_idx


  def getObjectiveValue(self, cur_sol):
    return np.sum([cur_sol[i]*self.costOfTest[i] for i in range(self.numTests)])

  def getRoundUpObjectiveValue(self, cur_sol):
    return np.sum([(1.0 if cur_sol[i] > 0 else 0) *self.costOfTest[i] for i in range(self.numTests)])

  def toString(self):
    out = ""
    out = f"Number of test: {self.numTests}\n"
    out+=f"Number of diseases: {self.numDiseases}\n"
    cst_str = " ".join([str(i) for i in self.costOfTest])
    out+=f"Cost of tests: {cst_str}\n"
    A_str = "\n".join([" ".join([str(j) for j in self.A[i]]) for i in range(0,self.A.shape[0])])
    out+=f"A:\n{A_str}"
    return out

File: src/solver/ip/IPInstance.java
package solver.ip;

import ilog.cplex.*;
import ilog.concert.*;

import java.util.*;

public class IPInstance implements Cloneable
{
    // IBM Ilog Cplex Solver 
    IloCplex cplex;

    int numTests;			// number of tests
    int numDiseases;		// number of diseases
    double[] costOfTest;  // [numTests] the cost of each test
    int[][] A;            // [numTests][numDiseases] 0/1 matrix if test is positive for disease
    double[] solution;
    Map<String, IloConstraint[]> constraintMap;
    IloNumVar[] isTestUsed;
    double objValue;
    int intObjValue = Integer.MAX_VALUE;
  
    public IPInstance()
    {
    super();
    }
  
    void init(int numTests, int numDiseases, double[] costOfTest, int[][] A) {
        assert(numTests >= 0) : "Init error: numtests should be non-negative " + numTests;
        assert(numDiseases >= 0) : "Init error: numtests should be non-negative " + numTests;
        assert(costOfTest != null) : "Init error: costOfTest cannot be null";
        assert(costOfTest.length == numTests) : "Init error: costOfTest length differ from numTests" + costOfTest.length + " vs. " + numTests;
        assert(A != null) : "Init error: A cannot be null";
        assert(A.length == numTests) : "Init error: Number of rows in A differ from numTests" + A.length + " vs. " + numTests;
        assert(A[0].length == numDiseases) : "Init error: Number of columns in A differ from numDiseases" + A[0].length + " vs. " + numDiseases;

        this.numTests = numTests;
        this.numDiseases = numDiseases;
        this.costOfTest = new double[numTests];
        for(int i=0; i < numTests; i++)
            this.costOfTest[i] = costOfTest[i];
        this.A = new int[numTests][numDiseases];
        for(int i=0; i < numTests; i++)
            for(int j=0; j < numDiseases; j++)
            this.A[i][j] = A[i][j];
        try {
            this.solution = this.solve();
        } catch (IloException e) {
            e.printStackTrace();
        }
        this.objValue = getObjectiveValue(this.solution);
        System.out.println("Solution: ");
        for (int i = 0; i < numTests; i++) {
            System.out.print(this.solution[i] + " ");
        }
        System.out.println();
        try {
            this.constraintMap = preComputeConstraints();
        } catch (IloException e) {
            e.printStackTrace();
        }
        // this.intObjValue = (int)Math.round(branchAndBoundBest());
        // System.out.println("Objective value: " + this.intObjValue);
        }

    public String toString() {
        StringBuffer buf = new StringBuffer();
        buf.append("Number of tests: " + numTests + "\n");
        buf.append("Number of diseases: " + numDiseases + "\n");
        buf.append("Cost of tests: " + Arrays.toString(costOfTest) + "\n");
        buf.append("A:\n");
        for(int i=0; i < numTests; i++)
            buf.append(Arrays.toString(A[i]) + "\n");
        return buf.toString();
    }

    public int[][] createDifferenceMatrix() {
        int[][] differenceMatrix = new int[numTests][numDiseases * (numDiseases - 1) / 2];

        int colIndex = 0;
        for (int j = 0; j < numDiseases; j++) {
            for (int k = j + 1; k < numDiseases; k++) {
                for (int i = 0; i < numTests; i++) {
                    differenceMatrix[i][colIndex] = A[i][j] ^ A[i][k]; // XOR operation
                }
                colIndex++;
            }
        }

        return differenceMatrix;
    }

    // Function to solve the IP using CPLEX
    public double[] solve() throws IloException 
    {
        this.cplex = new IloCplex();
        this.cplex.setOut(null);

        // Decision variables
        this.isTestUsed = this.cplex.numVarArray(numTests, 0.0, 1.0);

        // IloIntVar[] x = new IloIntVar[numTests];

        // for (int i = 0; i < numTests; i++) {
        //     x[i] = this.cplex.intVar(0, 1, "x" + i); // Integer variable bounded between 0 and 1
        // }

        // Objective function
        IloLinearNumExpr objective = this.cplex.linearNumExpr();
        for (int i = 0; i < numTests; i++) {
            objective.addTerm(costOfTest[i], isTestUsed[i]);
        }
        cplex.addMinimize(objective);

        // Differentiation constraints
        int[][] differenceMatrix = createDifferenceMatrix();
        for (int j = 0; j < differenceMatrix[0].length; j++) {
            IloLinearNumExpr constraint = this.cplex.linearNumExpr();
            for (int i = 0; i < numTests; i++) {
                constraint.addTerm(differenceMatrix[i][j], isTestUsed[i]);
            }
            this.cplex.addGe(constraint, 1);
        }

        // Solve the model
        this.cplex.solve();

        // Extract solution
        double[] solution = this.cplex.getValues(isTestUsed);
        return solution;
    }


    // Function to get the index of the greatest fractional value
    public int greatestFractionalIdx(double[] arr) {
        double maxFractional = 0;
        int maxIdx = -1;
        for (int i = 0; i < arr.length; i++) {
            double fractionalPart = Math.abs(arr[i] - (int) arr[i]);
            if (fractionalPart > maxFractional) {
                maxFractional = fractionalPart;
                maxIdx = i;
            }
        }
        return maxIdx;
    }

    public int distanceFromHalf(double[] arr) {
        double minDistance = 1;
        int maxIdx = -1;
        for (int i = 0; i < arr.length; i++) {
            double curDist = Math.abs(arr[i] - 0.5);
            if (curDist < minDistance) {
                minDistance = curDist;
                maxIdx = i;
            }
        }
        return maxIdx;
    }

    // Function to calculate the objective value
    public double getObjectiveValue(double[] curSol) {
        double objValue = 0;
        for (int i = 0; i < numTests; i++) {
            objValue += curSol[i] * costOfTest[i];
        }
        return objValue;
    }

    public double getRoundObjectiveValue(double[] curSol) {
        double objValue = 0;
        for (int i = 0; i < numTests; i++) {
            if (curSol[i] > 0) {
                objValue += costOfTest[i];
            }
        }
        return objValue;
    }

    private Map<String, IloConstraint[]> preComputeConstraints() throws IloException {
        Map<String, IloConstraint[]> constraintMap = new HashMap<>();

        for (int i = 0; i < numTests; i++) {
            IloConstraint[] constraints = new IloConstraint[2];
            constraints[0] = this.cplex.le(this.isTestUsed[i], 0);
            constraints[1] = this.cplex.ge(this.isTestUsed[i], 1);
            constraintMap.put("x_" + i, constraints);
        }

        return constraintMap;
    }
    
    public double[] applyConstraints(Map<Integer, Integer> constraints) throws IloException {
        List<IloConstraint> docplexConstraints = new ArrayList<>();

        for (Map.Entry<Integer, Integer> entry : constraints.entrySet()) {
            int key = entry.getKey();
            int val = entry.getValue();
            docplexConstraints.add(this.constraintMap.get("x_" + key)[val]);
        }
        for (IloConstraint constraint : docplexConstraints) {
            this.cplex.add(constraint);
        }

        this.cplex.solve();

        // double retVal = Double.POSITIVE_INFINITY;
        double[] newSol = new double[numTests];
        if (this.cplex.getStatus().equals(IloCplex.Status.Optimal)) {
            newSol = this.cplex.getValues(isTestUsed);
        }

        // Clear the constraints
        for (IloConstraint constraint : docplexConstraints) {
            this.cplex.remove(constraint);
        }

        // Return the solution and objective value
        // newSol and retVal
        return newSol;
    }

    public double branchAndBoundBest() {
        double[] bestSolution = new double[numTests];
        PriorityQueue<BnBNode> priorityQueue = new PriorityQueue<>();
        double upperBound = Double.POSITIVE_INFINITY;

        // Add initial node to priority queue
        priorityQueue.offer(new BnBNode(this.objValue, solution, new HashMap<>()));

        while (!priorityQueue.isEmpty()) {
            BnBNode currentNode = priorityQueue.poll();
            double newVal = currentNode.getValue();

            if (newVal > upperBound) {
                continue;
            }

            int branchIdx = greatestFractionalIdx(currentNode.getSolution());

            if (branchIdx == -1) {
                System.out.println("Integer solution found: " + newVal);
                if (newVal < upperBound) {
                    upperBound = newVal;
                    bestSolution = currentNode.getSolution();
                }
            } else {
                Map<Integer, Integer> constraints = currentNode.getConstraints();
                
                Map<Integer, Integer> newConstraints1 = new HashMap<>(constraints);
                newConstraints1.put(branchIdx, 1);
                double[] newSolution1 = {};
                try {
                    newSolution1 = applyConstraints(newConstraints1);
                } catch (IloException e) {
                    e.printStackTrace();
                }
                double newObjective1 = getObjectiveValue(newSolution1);
                double newRoundObjective1 = getRoundObjectiveValue(newSolution1);
                upperBound = Math.min(upperBound, newRoundObjective1);
                if (newObjective1 <= upperBound) {
                    priorityQueue.offer(new BnBNode(newObjective1, newSolution1, newConstraints1));
                }

                Map<Integer, Integer> newConstraints0 = new HashMap<>(constraints);
                newConstraints0.put(branchIdx, 0);
                double[] newSolution0 = {};
                try {
                    newSolution0 = applyConstraints(newConstraints0);
                } catch (IloException e) {
                    e.printStackTrace();
                }
                double newObjective0 = getObjectiveValue(newSolution0);
                double newRoundObjective0 = getRoundObjectiveValue(newSolution0);
                upperBound = Math.min(upperBound, newRoundObjective0);
                if (newObjective0 <= upperBound) {
                    priorityQueue.offer(new BnBNode(newObjective0, newSolution0, newConstraints0));
                }
            }
        }

        System.out.println("Best solution: ");
        for (int i = 0; i < numTests; i++) {
            System.out.print(bestSolution[i] + " ");
        }
        System.out.println();

        return upperBound;
    }

    @Override
    public IPInstance clone() {
        IPInstance clone = new IPInstance();
        clone.numTests = this.numTests;
        clone.numDiseases = this.numDiseases;
        clone.costOfTest = this.costOfTest.clone();
        clone.A = new int[numTests][numDiseases];
        for (int i = 0; i < numTests; i++) {
            clone.A[i] = this.A[i].clone();
        }
        try {
            clone.solution = clone.solve();
        } catch (IloException e) {
            e.printStackTrace();
        }
        clone.objValue = getObjectiveValue(clone.solution);
        // System.out.println("Solution: ");
        // for (int i = 0; i < numTests; i++) {
        //     System.out.print(this.solution[i] + " ");
        // }
        // System.out.println();
        try {
            clone.constraintMap = clone.preComputeConstraints();
        } catch (IloException e) {
            e.printStackTrace();
        }
        return clone;
    }

}